src/shared/components/query/quickSearch/QuickSearch.tsx

```diff
@@ -84,76 +85,93 @@ export default class QuickSearch extends React.Component {
-                client.getAllGenesUsingGETWithHttpInfo({keyword: input, pageSize: DEFAULT_PAGE_SIZE + (SHOW_MORE_SIZE * this.genePageMultiplier)}),
-                client.getAllGenesUsingGETWithHttpInfo({keyword: input, projection: "META"}),
-                client.getAllPatientsUsingGETWithHttpInfo({keyword: input, pageSize: DEFAULT_PAGE_SIZE + (SHOW_MORE_SIZE * this.patientPageMultiplier), projection: "DETAILED"}),
-                client.getAllPatientsUsingGETWithHttpInfo({keyword: input, projection: "META"}),
-                // we use sleep method because if the response is cached by superagent, react-select can't render the options for some reason
-                sleep(0)]).then(async (response: any) => {
-                let studyOptions: any = response[0].body.map(this.studyToOption);
-
-                let studyCount = {
-                    value: parseInt(response[1].headers["total-count"]) - studyOptions.length,
-                    type: OptionType.STUDY_COUNT
-                }
-                const geneOptions: any = response[2].body.map(this.geneToOption);
-                const geneCount = {
-                    value: parseInt(response[3].headers["total-count"]) - geneOptions.length,
-                    type: OptionType.GENE_COUNT
-                }
-                const patientOptions: any = response[4].body.map(this.patientToOption);
-                const patientCount = {
-                    value: parseInt(response[5].headers["total-count"]) - patientOptions.length,
-                    type: OptionType.PATIENT_COUNT
-                }
+    // tslint:disable-next-line:member-ordering
+    private geneStudyQuery = remoteData<GeneStudyQuery>(()=>{
+        return Promise.resolve({
+            type: GeneStudyQueryType.STUDY_LIST,
+            query: AppConfig.serverConfig.skin_quick_search_gene_query_cancer_study_list!,
+            name: "TCGA PanCancer Atlas studies"
+        });
+    });
 
-                if (((geneOptions.length + patientOptions.length) < (2 * DEFAULT_PAGE_SIZE)) && studyCount.value > 0) {
-                    const spillover = (2 * DEFAULT_PAGE_SIZE) - (geneOptions.length + patientOptions.length);
-                    await client.getAllStudiesUsingGETWithHttpInfo({keyword: input,
-                        pageSize: spillover + DEFAULT_PAGE_SIZE + (SHOW_MORE_SIZE * this.studyPageMultiplier)})
-                        .then((response: any) => {
-                            studyOptions = response.body.map(this.studyToOption);
-                        });
-                    studyCount.value = studyCount.value - spillover;
-                }
+    // tslint:disable-next-line:member-ordering
+    private options = remoteData<any[]>({
+        await: () => [
+            this.geneStudyQuery
+        ],
+        invoke: () => {
+
+            const input = this.inputValue;
+            const geneStudyQuery = this.geneStudyQuery.result;
+
+            if (input.length > 0) {
```

Comment:
@dippindots i know you didn't write this code but we should clean this up a little.   very hard to read flow.  we should put all the get calls that in a separately declared array and then use await on Promise.All and then divide up the rest of this routine into utility functions. lets discuss.  we can do this in a follow up PR though.

Reply:
@alisman Sure!